### PR TITLE
Handle miner fees for NG

### PR DESCRIPTION
**Pull request by gorillainduction**
* Split the fees by generation
* Store fees in the block state
* Handle delay of miner rewards

There are certainly more optimizations that can be done here, and some more testing.

### CODE REVIEW

File: apps/aecore/src/aec_block_key_candidate.erl [hunk 1]
```diff
@@ -6,47 +6,28 @@
...
--opaque block_info() :: #{trees := aec_trees:trees(),
-                          fees := fees_info(),
-                          txs_tree := aec_txs_trees:txs_tree(),
-                          adj_chain := [aec_headers:header()]}.
-
 %% -- API functions ----------------------------------------------------------
--spec create(aec_blocks:block() | aec_blocks:block_header_hash(), fees_info()) ->
-        {ok, aec_blocks:block(), block_info()} | {error, term()}.
-create(BlockHash, FeesInfo) when is_binary(BlockHash) ->
+-spec create(aec_blocks:block() | aec_blocks:block_header_hash()) ->
```
Comment (hanssv) [inline apps/aecore/src/aec_block_key_candidate.erl]: This is legacy from pre-NG block creation it should be enough with the version that takes a hash... The other variant is only used in tests!??

File: apps/aecore/src/aec_block_key_candidate.erl [hunk 2]
```diff
@@ -66,44 +47,38 @@ adjust_target(Block, AdjHeaders) ->
...
             Error
     end.
 
-int_create(PrevBlockHash, PrevBlock, FeesInfo, Trees, Miner, AdjChain) ->
-    {ok, Block, BlockInfo} = int_create_block(PrevBlockHash, PrevBlock, FeesInfo, Trees, Miner),
+int_create(PrevBlockHash, PrevBlock, Miner, AdjChain) ->
+    {ok, Trees} = aec_chain_state:calculate_keyblock_state(PrevBlockHash, Miner),
+    Block = int_create_block(PrevBlockHash, PrevBlock, Miner, Trees),
     case adjust_target(Block, AdjChain) of
-        {ok, AdjBlock} -> {ok, AdjBlock, BlockInfo#{ adj_chain => AdjChain }};
```
Comment (hanssv) [inline apps/aecore/src/aec_block_key_candidate.erl]: Premature optimization - we will need that adjacent chain when implementing an optimized update for keyblocks...

Easy to re-introduce though.

File: apps/aecore/src/aec_block_key_candidate.erl [hunk 3]
```diff
@@ -66,44 +47,38 @@ adjust_target(Block, AdjHeaders) ->
...
-            int_create(BlockHash, Block, FeesInfo, Trees, Miner, AdjChain);
+            int_create(BlockHash, Block, Miner, AdjChain);
         {error, _} = Error ->
             Error
     end.
 
-int_create(PrevBlockHash, PrevBlock, FeesInfo, Trees, Miner, AdjChain) ->
-    {ok, Block, BlockInfo} = int_create_block(PrevBlockHash, PrevBlock, FeesInfo, Trees, Miner),
+int_create(PrevBlockHash, PrevBlock, Miner, AdjChain) ->
+    {ok, Trees} = aec_chain_state:calculate_keyblock_state(PrevBlockHash, Miner),
```
Comment (hanssv) [inline apps/aecore/src/aec_block_key_candidate.erl]: Aren't we doing the job twice here? We have already performed `aec_chain_state:insert_block` for `PrevBlock`?

File: apps/aecore/test/aec_chain_tests.erl [hunk 4]
```diff
@@ -734,6 +734,7 @@ fork_get_transaction() ->
     RecipientPubKey = <<42:32/unit:8>>,
     PresetAccounts = [{SenderPubKey, 100}],
     meck:expect(aec_genesis_block_settings, preset_accounts, 0, PresetAccounts),
+    meck:expect(aec_governance, miner_reward_delay, 0, 100),
```
Comment (hanssv) [inline apps/aecore/test/aec_chain_tests.erl]: 😛

File: apps/aecore/src/aec_block_key_candidate.erl (hunk 3 above)
Comment (sennui) [reply apps/aecore/src/aec_block_key_candidate.erl]: yeah, that's good it's shared with aec chain state, tho, so easier to refactor

Comment (gorillainduction) [reply apps/aecore/src/aec_block_key_candidate.erl]: You are misreading. The `calculate_keykblock_state` is calculating what the state is if there is a keyblock added on top of the hash. Should be clarified in the code, and possibly have a better name.

Anyway, the job is done once (here) and if the block is mined, it is performed again with the actual block.

File: apps/aecore/src/aec_block_key_candidate.erl (hunk 2 above)
Comment (gorillainduction) [reply apps/aecore/src/aec_block_key_candidate.erl]: Please reintroduce it if needed. Currently it is not used.

File: apps/aecore/test/aec_chain_tests.erl (hunk 4 above)
Comment (gorillainduction) [reply apps/aecore/test/aec_chain_tests.erl]: :man_shrugging:

**Pull request merged**